UnitTests/GitUITests/CancellationTokenSequenceTests.cs

```diff
@@ -0,0 +1,163 @@
+
+            Assert.True(token.IsCancellationRequested);
+        }
+
+        [Test]
+        public void Cancel_is_idempotent()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            var token = sequence.Next();
+
+            Assert.False(token.IsCancellationRequested);
+
+            sequence.CancelCurrent();
+            sequence.CancelCurrent();
+            sequence.CancelCurrent();
+            sequence.CancelCurrent();
+        }
+
+        [Test]
+        public void Cancel_does_not_throw_if_no_token_yet_issued()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.CancelCurrent();
+        }
+
+        [Test]
+        public void Dispose_cancels_previous_token()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            var token = sequence.Next();
+
+            Assert.False(token.IsCancellationRequested);
+
+            sequence.Dispose();
+
+            Assert.True(token.IsCancellationRequested);
+        }
+
+        [Test]
+        public void Dispose_is_idempotent()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.Next();
+
+            sequence.Dispose();
+            sequence.Dispose();
+            sequence.Dispose();
+            sequence.Dispose();
+        }
+
+        [Test]
+        public void Dispose_does_not_throw_if_no_token_yet_issued()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.Dispose();
+        }
+
+        [Test]
+        public void Next_handles_concurrent_requests_correctly()
+        {
+            const int loopCount = 10000;
+
+            // Assume hyperthreading, so halve logical processors (could use WMI or P/Invoke for more robust answer)
+            var coreCount = Math.Max(1, Environment.ProcessorCount / 2);
+
+            if (coreCount == 1)
+            {
+                Assert.Inconclusive("This test requires more than one processor to run.");
+            }
+
+            using (var sequence = new CancellationTokenSequence())
+            using (var barrier = new Barrier(coreCount))
+            using (var countdown = new CountdownEvent(loopCount * coreCount))
+            {
+                var completedCount = 0;
+
+                var completionTokenSource = new CancellationTokenSource();
+                var completionToken = completionTokenSource.Token;
+
+                for (var i = 0; i < coreCount; i++)
+                {
+                    new Thread(ThreadMethod).Start();
+                }
+
+                Assert.True(countdown.Wait(TimeSpan.FromSeconds(10)), "Test should have completed within a reasonable amount of time");
+
+                Assert.AreEqual(loopCount, completedCount);
+
+                void ThreadMethod()
+                {
+                    while (true)
+                    {
+                        barrier.SignalAndWait(completionToken);
```

Comment:
I don't know if it is because debugger stops on unhandled exceptions, but the test got stuck here if  I press to continue. Perhaps it is, because the assertion in the 135 line did not fail - killed the test after a minute.
To not be bothered by the OperationCanceledException I suggest:

```csharp
                        try
                        {
                            barrier.SignalAndWait(completionToken);
                        }
                        catch (OperationCanceledException)
                        {
                            break;
                        }

```

Reply:
I'll see if I can reproduce this. Do you know where code was blocked? The only way I can think that it would not terminate is if the countdown doesn't reach zero for some reason.

Reply:
Both threads were blocked on `barrier.SignalAndWait(completionToken);`

Reply:
Or just this?

```csharp
barrier.SignalAndWait();
if (completionToken.IsCancellationRequested)
{
  return;
}
```

Reply:
@sharwell I think that is a better solution than using exceptions for control flow 👍